spec/requests/orders_spec.rb

```diff
@@ -0,0 +1,143 @@
+require "rails_helper"
+
+RSpec.describe OrdersController, type: :controller do
+  include ActiveJob::TestHelper
+
+  let(:user) { create :user }
+  let!(:product) { create(:product, status: "enabled") }
+
+  before do
+    sign_in user
+  end
+
+  describe "GET #index" do
+    let!(:order) { create :order, user: user, status: status }
+    
+    context "when get all orders success" do
+      let(:status) { :wait_confirm }
+
+      before { get :index }
+
+      it do
+        expect(response).to be_successful
+        expect(response).to render_template "index"
+        expect(assigns[:orders]).to eq [order]
+      end
+    end
+
+    context "when get orders by status" do
+      let(:status) { :delivering }
+
+      before { get :index, params: { status: status} }
+
+      it do
+        expect(response).to be_successful
+        expect(response).to render_template "index"
+        expect(assigns[:orders]).to eq [order]
+        expect(assigns[:orders].first.status).to eq status.to_s
+      end
+    end
+  end
+
+  describe "POST #create" do
+    subject { post :create, params: { order: order_params } }
+
+    let(:cart_items) { { product.id => 1 }.to_json }
+    let(:order_params) { { address: "123 Main St", phone: "123-456-7890" } }
+
+    before { request.cookies[:cart_items] = cart_items }
+
+    after { clear_enqueued_jobs }
+
+    describe "callbacks" do
+      describe "#cart_items" do
+        it do
+          expect { subject }.to change { assigns(:cart_items) }
+                    .from(assigns(:cart_items)).to(JSON.parse(cart_items))
+        end
+      end
+    end
+
+    context "when valid product and order params" do
+      let(:order_mailer_jobs) do
+        ActiveJob::Base.queue_adapter.enqueued_jobs.select do |job|
+          job[:args][0] == "OrderMailer" && job[:args][1] == "place_order"
+        end
+      end
+
+      it do
+        expect { subject }.to change(Order, :count).by(1)
+        expect(response).to redirect_to orders_path
+        expect(response.cookies).to include("cart_items" => nil)
+        expect(flash[:success]).to eq "Order successfully placed! Please check your email to track the notification"
+        expect(order_mailer_jobs.size).to eq 1
+      end
```

Comment:
Đây 1 lý do a tách 1 it 1 expect, ở đây em k thể để expect { subject } 2 lần được.
tách thành nhiều context nha

```ruby
context "when response" do
  before { subject }
  
  it { expect(response).to redirect_to orders_path }
end

context "when create new order" do
  it {expect { subject }.to change(Order, :count).by(1) }
end

context "when enqueued_jobs" do
# có matcher để mình check job được enqueue hay không
# https://www.rubydoc.info/gems/rspec-rails/RSpec%2FRails%2FMatchers:have_enqueued_job
end
```

Reply:
Dạ okie anh để em tách ra ạ